File: Samples/HtmlToPdf/XmlToPdfUsingXslt/C#/XmlToPdfUsingXslt.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Threading.Tasks;
using System.Xml;
using System.Xml.Xsl;

using BitMiracle.Docotic;
using BitMiracle.Docotic.Pdf.HtmlToPdf;

namespace XmlToPdfUsingXslt;

class XmlToPdfUsingXslt
{
    static async Task Main()
    {
        // NOTE:
        // Without a license, the library won't allow you to create or read PDF documents.
        // To get a free time-limited license key, use the form on
        // https://bitmiracle.com/pdf-library/download

        LicenseManager.AddLicenseData("PUT-LICENSE-HERE");

        var transform = CreateTransform(@"..\Sample Data\invoice.xsl");
        var html = TransformToHtml(@"..\Sample Data\invoice.xml", transform);

        var outputFileName = "XmlToPdfUsingXslt.pdf";
        using (var converter = await HtmlConverter.CreateAsync())
        {
            // If the converted HTML uses relative paths for images, CSS files, or other resources,
            // don't forget to specify the base URL.
            using var pdf = await converter.CreatePdfFromStringAsync(html);
            pdf.Save(outputFileName);
        }

        Console.WriteLine($"The output is located in {Environment.CurrentDirectory}");

        Process.Start(new ProcessStartInfo(outputFileName) { UseShellExecute = true });
    }

    private static XslCompiledTransform CreateTransform(string xsltFileName)
    {
        var transform = new XslCompiledTransform();
        transform.Load(xsltFileName);
        return transform;
    }

    private static string TransformToHtml(string xmlFileName, XslCompiledTransform transform)
    {
        using var xmlReader = XmlReader.Create(xmlFileName);
        using var sw = new StringWriter();
        using var writer = XmlWriter.Create(sw, transform.OutputSettings);
        transform.Transform(xmlReader, null, writer);
        return sw.ToString();
    }
}

File: Samples/HtmlToPdf/XmlToPdfUsingXslt/VB.NET/XmlToPdfUsingXslt/XmlToPdfUsingXslt.vb
Imports System.IO
Imports System.Xml
Imports System.Xml.Xsl
Imports BitMiracle.Docotic.Pdf.HtmlToPdf

Module XmlToPdfUsingXslt
    Public Sub Main()
        Task.Run(
            Async Function()
                Await XmlToPdfUsingXslt()
            End Function
        ).GetAwaiter().GetResult()
    End Sub

    Private Async Function XmlToPdfUsingXslt() As Task
        ' NOTE:
        ' Without a license, the library won't allow you to create or read PDF documents.
        ' To get a free time-limited license key, use the form on
        ' https://bitmiracle.com/pdf-library/download

        LicenseManager.AddLicenseData("PUT-LICENSE-HERE")

        Dim transform = CreateTransform("..\Sample Data\invoice.xsl")
        Dim html = TransformToHtml("..\Sample Data\invoice.xml", transform)

        Dim outputFileName = "XmlToPdfUsingXslt.pdf"

        Using converter = Await HtmlConverter.CreateAsync()
            ' If the converted HTML uses relative paths for images, CSS files, or other resources,
            ' don't forget to specify the base URL.
            Using pdf = Await converter.CreatePdfFromStringAsync(html)
                pdf.Save(outputFileName)
            End Using
        End Using

        Console.WriteLine($"The output is located in {Environment.CurrentDirectory}")

        Process.Start(New ProcessStartInfo(outputFileName) With {.UseShellExecute = True})
    End Function

    Private Function CreateTransform(xsltFileName As String) As XslCompiledTransform
        Dim transform = New XslCompiledTransform()
        transform.Load(xsltFileName)
        Return transform
    End Function

    Private Function TransformToHtml(xmlFileName As String, transform As XslCompiledTransform) As String
        Using reader = XmlReader.Create(xmlFileName)
            Using sw As New StringWriter()
                Using writer = XmlWriter.Create(sw, transform.OutputSettings)
                    transform.Transform(reader, Nothing, writer)
                End Using
                Return sw.ToString()
            End Using
        End Using
    End Function
End Module
